### PR TITLE
[SPARK-17569] Make StructuredStreaming FileStreamSource batch generation faster

### PR DESCRIPTION
**Pull request by brkyvz**
## What changes were proposed in this pull request?

While getting the batch for a `FileStreamSource` in StructuredStreaming, we know which files we must take specifically. We already have verified that they exist, and have committed them to a metadata log. When creating the FileSourceRelation however for an incremental execution, the code checks the existence of every single file once again!

When you have 100,000s of files in a folder, creating the first batch takes 2 hours+ when working with S3! This PR disables that check
## How was this patch tested?

Added a unit test to `FileStreamSource`.

### CODE REVIEW

**Comment by brkyvz** [comment]
cc @zsxwing @yhuai

**Comment by yhuai** [comment]
also cc @cloud-fan

**Comment by petermaxlee** [comment]
Can you test this by deleting the file on purpose, and see what kind of exceptions are thrown?

**Comment by petermaxlee** [comment]
I looked into this. I think there are two ways that you can intercept any calls to HDFS.

The first way is slightly hacky but pretty simple. FileSystem.addFileSystemForTesting is a package private method that can be used to inject a mock file system. You can create an implementation of FilterFileSystem and pass it in as "file" schema. Then all accesses to local file system will go through your implementation. Of course, you can also use a mocking library to do that, but that is not as clean since FilterFileSystem is a public interface.

The second way is more robust and does not depend on any private APIs. Create an implementation of FilterFileSystem by pointing to LocalFileSystem, e.g. call it MockFileSystem. MockFileSystem.getScheme should return "mockfs://". You can then use this as the path when passing to structured streaming. This is probably a more robust, generic solution.

**Comment by yhuai** [comment]
@petermaxlee I believe you will get a runtime exception saying that the file does not exist. 

Also, regarding your options 2, are you suggesting that users of structured streaming to use such a mock fs? Or you are suggesting that structured streaming to use such a fs. Also, why LocalFS is related to this case?

**Comment by brkyvz** [comment]
@yhuai The suggestions are for purely testing purposes, to make sure that StructuredStreaming doesn't check for file existence twice.

**Comment by brkyvz** [comment]
@petermaxlee Thank you for the suggestions for testing. I will try out Option 1, since 2 is a bit much work for a minor PR as this.

**Comment by yhuai** [comment]
ok, got it. Thanks!

**Comment by brkyvz** [comment]
Added test using Option 2 in the end.

**Comment by zsxwing** [comment]
LGTM. Let's run the test again since there are several PRs got merged in the past two days.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/FileStreamSourceSuite.scala [hunk 1]
```diff
@@ -73,4 +83,44 @@ class FileStreamSourceSuite extends SparkFunSuite {
...
   }
 
+  testWithUninterruptibleThread("do not recheck that files exist during getBatch") {
+    withTempDir { temp =>
+      spark.conf.set(
+        s"fs.$scheme.impl",
+        classOf[ExistsThrowsExceptionFileSystem].getName)
+      // add the metadata entries as a pre-req
+      val dir = new File(temp, "dir") // use non-existent directory to test whether log make the dir
+      val metadataLog = new HDFSMetadataLog[Array[FileEntry]](spark, dir.getAbsolutePath)
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/FileStreamSourceSuite.scala]: Need to use `FileStreamSourceLog` here to add FileEntry.

**Comment by brkyvz** [comment]
@zsxwing Thanks for the comment. Updated!

**Comment by zsxwing** [comment]
LGTM. Merging to master and 2.0. Thanks!

**Comment by zsxwing** [comment]
There are some conflicts with 2.0. Could you submit a patch for branch 2.0?

**Pull request closed without merging**